File: src/dere-plugins/dere_plugin/skills/recalling-past-context/scripts/related_entities.py
#!/usr/bin/env python3
"""Get related entities."""

from __future__ import annotations

import json
import sys
from pathlib import Path

import requests

sys.path.insert(0, str(Path(__file__).parent.parent.parent / "scripts"))
from config_reader import get_daemon_url


def get_related_entities(entity_name: str, limit: int = 10) -> dict | None:
    """Get entities related to given entity."""
    daemon_url = get_daemon_url()

    try:
        response = requests.get(
            f"{daemon_url}/entities/related/{entity_name}", params={"limit": limit}, timeout=5
        )
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"Error getting related entities: {e}", file=sys.stderr)
        return None


def main():
    """Main entry point."""
    if len(sys.argv) < 2:
        print("Usage: related_entities.py <entity_name> [limit]", file=sys.stderr)
        sys.exit(1)

    entity_name = sys.argv[1]
    limit = int(sys.argv[2]) if len(sys.argv) > 2 else 10

    result = get_related_entities(entity_name, limit)

    if result:
        print(json.dumps(result, indent=2))
    else:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/dere_plugins/dere_core/skills/recall/scripts/related_entities.ts
import { getDaemonUrl } from "../../../scripts/config_reader.js";

async function getRelatedEntities(
  entityName: string,
  limit = 10,
): Promise<Record<string, unknown> | null> {
  const daemonUrl = await getDaemonUrl();
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 5_000);
  try {
    const url = new URL(`${daemonUrl}/kg/entity/${entityName}/related`);
    url.searchParams.set("limit", String(limit));
    const response = await fetch(url.toString(), {
      method: "GET",
      signal: controller.signal,
    });
    if (!response.ok) {
      console.error(`Error getting related entities: ${response.status} ${response.statusText}`);
      return null;
    }
    return (await response.json()) as Record<string, unknown>;
  } catch (error) {
    console.error(`Error getting related entities: ${String(error)}`);
    return null;
  } finally {
    clearTimeout(timeout);
  }
}

async function main(): Promise<void> {
  const entityName = process.argv[2];
  if (!entityName) {
    console.error("Usage: related_entities.ts <entity_name> [limit]");
    process.exit(1);
  }

  const limitArg = process.argv[3];
  const limit = limitArg ? Number.parseInt(limitArg, 10) : 10;

  const result = await getRelatedEntities(entityName, Number.isFinite(limit) ? limit : 10);
  if (result) {
    console.log(JSON.stringify(result, null, 2));
  } else {
    process.exit(1);
  }
}

if (import.meta.main) {
  void main();
}
